core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java

```diff
@@ -371,7 +372,7 @@ protected PlanNodeStatsEstimate visitComparisonExpression(ComparisonExpression n
             }
 
             SymbolStatsEstimate leftStats = getExpressionStats(left);
-            Optional<Symbol> leftSymbol = left instanceof SymbolReference ? Optional.of(Symbol.from(left)) : Optional.empty();
+            Optional<Symbol> leftSymbol = getSymbolReference(left);
```

Comment:
I'm not sure it's right thing to do. For example `io.trino.cost.ComparisonStatsCalculator#estimateFilterRange` will set column statistics for that symbol based on cast `leftStats` (please correct me if I'm wrong). It could be that `CAST` is narrowing down, so that target domain stats don't make sense for original symbol.

If this code brings benefits, then I would do either:
* add TODO to `getSymbolReference` to somehow limit unwrapping of `CAST` to cases where target domains is wider or equal
* make `getSymbolReference` unwrap `CAST` to to cases where target domains is wider or equal

Reply:
I've now added logic to do this unwrapping only if the cast is `isTypeOnlyCoercion`.
I'm assuming this will ensure that the CAST is to a wider domain

Reply:
> I've now added logic to do this unwrapping only if the cast is isTypeOnlyCoercion.

`isTypeOnlyCoercion`  is about Trino stack representation.
This is not the same as "stats representation", the two meanings should not be conflated.
At the very least you should have a 
- a method in `StatsUtil` that tells which type pairs (which casts)  it's safe to unwrap
- have a test that verifies `StatsUtil.toStatsRepresentation` is the same before and after the cast, for the types allowed for unwrapping

Reply:
I've added `StatsUtil#areStatsRepresentationsCompatible` now and some testing around it in `FilterStatsCalculator#testUnwrapCastSymbolReference`